d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/RelativeStateUpdater.java

```diff
@@ -0,0 +1,217 @@
+
+    // Remove the trackerClients from original map if there is any change in uri list
+    if (maybeClusterGenerationId.isPresent())
+    {
+      for (TrackerClient trackerClient : trackerClientStateMap.keySet())
+      {
+        if (!trackerClients.contains(trackerClient))
+        {
+          trackerClientStateMap.remove(trackerClient);
+          partitionLoadBalancerState.getQuarantineMap().remove(trackerClient);
+          partitionLoadBalancerState.getQuarantineHistory().remove(trackerClient);
+          partitionLoadBalancerState.getRecoveryMap().remove(trackerClient);
+        }
+      }
+    }
+
+    // Calculate the base health score before we override them when handling the quarantine and recovery
+    calculateBaseHealthScore(trackerClients, partitionId);
+
+    // Update cluster generation id if it's changed
+    if (maybeClusterGenerationId.isPresent())
+    {
+      _partitionLoadBalancerStateMap.get(partitionId).setClusterGenerationId(maybeClusterGenerationId.get());
+    }
+  }
+
+  private void calculateBaseHealthScore(Set<TrackerClient> trackerClients, int partitionId)
+  {
+    // Snap stats for each tracker client, we want to get snap the stats because they can change any time during we calculate the new health score
+    long sumAvgLatency = 0;
+    PartitionLoadBalancerState partitionLoadBalancerState = _partitionLoadBalancerStateMap.get(partitionId);
+    ConcurrentMap<TrackerClient, TrackerClientState> trackerClientStateMap = partitionLoadBalancerState.getTrackerClientStateMap();
+    Map<TrackerClient, CallTracker.CallStats> latestCallStatsMap = new HashMap<>();
+
+    for (TrackerClient trackerClient : trackerClients)
+    {
+      CallTracker.CallStats latestCallStats = trackerClient.getLatestCallStats();
+      latestCallStatsMap.put(trackerClient, latestCallStats);
+
+      sumAvgLatency += Math.round(latestCallStats.getCallTimeStats().getAverage());
+    }
+    long clusterAvgLatency = sumAvgLatency / trackerClients.size();
+    partitionLoadBalancerState.setClusterAvgLatency(clusterAvgLatency);
+
+    // Update health score
+    for (TrackerClient trackerClient : trackerClients)
+    {
+      CallTracker.CallStats latestCallStats = latestCallStatsMap.get(trackerClient);
+
+      if (trackerClientStateMap.containsKey(trackerClient))
+      {
+        TrackerClientState trackerClientState = trackerClientStateMap.get(trackerClient);
+        int callCount = latestCallStats.getCallCount();
+        double errorRate = TrackerClientState.getErrorRateByType(latestCallStats.getErrorTypeCounts(), callCount);
+        long latency = Math.round(latestCallStats.getCallTimeStats().getAverage());
+        // If it is an existing tracker client
+        double oldHealthScore = trackerClientState.getHealthScore();
+        double newHealthScore = oldHealthScore;
+        if (TrackerClientState.isUnhealthy(trackerClientState, clusterAvgLatency, callCount, latency, errorRate,
```

Comment:
I'd rather keep decision making in this class instead of TrackerClientState since this is very similar to what Degrader was doing.